ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,895 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+			rt = ext->external_info;
+			if (!rt)
+				continue;
+
+			for (rn = route_top(rt); rn; rn = route_next(rn))
+				if (rn->info) {
+					ei = rn->info;
+					if (is_prefix_default(&ei->p))
+						continue;
+
+					/* Check the AS-external-LSA
+					 * should be originated.
+					 */
+					if (!ospf_redistribute_check(ospf, ei,
+								     NULL))
+						continue;
+
+					aggr = ospf_external_aggr_match(ospf,
+									&ei->p);
+
+					/* If matching aggregator found, Add
+					 * the external route reference to the
+					 * aggregator and originate the aggr
+					 * route if it is advertisable.
+					 * flush the external LSA if it is
+					 * already originated for this external
+					 * prefix.
+					 */
+					if (aggr) {
+						ospf_originate_summary_lsa(
+							ospf, aggr, ei);
+
+						/* All aggregated external rts
+						 * are handled here.
+						 */
+						ospf_handle_aggregated_exnl_rt(
+							ospf, aggr, ei);
+						continue;
+					}
+
+					/* External routes which are only out
+					 * of aggregation will be handled here.
+					 */
+					ospf_handle_exnl_rt_after_aggr_del(ospf,
+									   ei);
+				}
+		}
+	}
+}
+
+static void
+ospf_aggr_handle_advertise_change(struct ospf *ospf,
+				  struct ospf_external_aggr_rt *aggr,
+				  struct external_info *ei_aggr)
+{
+	struct ospf_lsa *lsa;
+
+	/* Check if advertise option modified. */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_NO_ADVERTISE)) {
+
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: Don't originate the summary address,It is configured to not-advertise.",
+				__func__);
+
+		if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: No-advertise,So Flush the Aggregate route(%s/%d)",
+					__func__, inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			ospf_external_lsa_flush(ospf, 0, &aggr->p, 0);
+
+			UNSET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		}
+		return;
+	}
+
+	if (!CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Now it is advatisable", __func__);
+
+		lsa = ospf_external_info_find_lsa(ospf, &ei_aggr->p);
+		if (lsa && IS_LSA_MAXAGE(lsa)) {
+			/* This is special case.
+			 * If a summary route need to be originated but where
+			 * summary route already exist in lsdb with maxage, then
+			 * it need to be refreshed.
+			 */
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: It is already with Maxage, So refresh it (%s/%d)",
+					__func__, inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			ospf_external_lsa_refresh(ospf, lsa, ei_aggr,
+						  LSA_REFRESH_FORCE, 1);
+
+			SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+
+		} else {
+
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: Originate Aggregate LSA (%s/%d)",
+					__func__, inet_ntoa(aggr->p.prefix),
```

Comment:
use `%pI4` to print ipv4 addresses rather than `inet_ntoa`

Reply:
done